File: src/main/java/org/gradle/api/plugins/shadow/relocation/SimpleRelocator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.gradle.api.plugins.shadow.relocation;

import org.gradle.mvn3.org.codehaus.plexus.util.SelectorUtils;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * @author Jason van Zyl
 * @author Mauro Talevi
 *
 * Modified from org.apache.maven.plugins.shade.relocation.SimpleRelocator.java
 *
 * Modifications
 * @author John Engelman
 */
public class SimpleRelocator implements Relocator {

    private final String pattern;

    private final String pathPattern;

    private final String shadedPattern;

    private final String shadedPathPattern;

    private final Set<String> includes;

    private final Set<String> excludes;

    private final boolean rawString;

    public SimpleRelocator(String patt, String shadedPattern, List<String> includes, List<String> excludes) {
        this(patt, shadedPattern, includes, excludes, false);
    }

    public SimpleRelocator(String patt, String shadedPattern, List<String> includes, List<String> excludes,
                           boolean rawString) {
        this.rawString = rawString;

        if (rawString) {
            this.pathPattern = patt;
            this.shadedPathPattern = shadedPattern;

            this.pattern = null; // not used for raw string relocator
            this.shadedPattern = null; // not used for raw string relocator
        } else {
            if (patt == null) {
                this.pattern = "";
                this.pathPattern = "";
            } else {
                this.pattern = patt.replace('/', '.');
                this.pathPattern = patt.replace('.', '/');
            }

            if (shadedPattern != null) {
                this.shadedPattern = shadedPattern.replace('/', '.');
                this.shadedPathPattern = shadedPattern.replace('.', '/');
            } else {
                this.shadedPattern = "hidden." + this.pattern;
                this.shadedPathPattern = "hidden/" + this.pathPattern;
            }
        }

        this.includes = normalizePatterns(includes);
        this.excludes = normalizePatterns(excludes);
    }

    private static Set<String> normalizePatterns(Collection<String> patterns) {
        Set<String> normalized = null;

        if (patterns != null && !patterns.isEmpty()) {
            normalized = new LinkedHashSet<String>();

            for (String pattern : patterns) {

                String classPattern = pattern.replace('.', '/');

                normalized.add(classPattern);

                if (classPattern.endsWith("/*")) {
                    String packagePattern = classPattern.substring(0, classPattern.lastIndexOf('/'));
                    normalized.add(packagePattern);
                }
            }
        }

        return normalized;
    }

    private boolean isIncluded(String path) {
        if (includes != null && !includes.isEmpty()) {
            for (String include : includes) {
                if (SelectorUtils.matchPath(include, path, true)) {
                    return true;
                }
            }
            return false;
        }
        return true;
    }

    private boolean isExcluded(String path) {
        if (excludes != null && !excludes.isEmpty()) {
            for (String exclude : excludes) {
                if (SelectorUtils.matchPath(exclude, path, true)) {
                    return true;
                }
            }
        }
        return false;
    }

    public boolean canRelocatePath(String path) {
        if (rawString) {
            return Pattern.compile(pathPattern).matcher(path).find();
        }

        if (path.endsWith(".class")) {
            path = path.substring(0, path.length() - 6);
        }

        if (!isIncluded(path) || isExcluded(path)) {
            return false;
        }

        // Allow for annoying option of an extra / on the front of a path. See MSHADE-119; comes from getClass().getResource("/a/b/c.properties").
        return path.startsWith(pathPattern) || path.startsWith("/" + pathPattern);
    }

    public boolean canRelocateClass(String clazz) {
        return !rawString && clazz.indexOf('/') < 0 && canRelocatePath(clazz.replace('.', '/'));
    }

    public String relocatePath(String path) {
        if (rawString) {
            return path.replaceAll(pathPattern, shadedPathPattern);
        } else {
            return path.replaceFirst(pathPattern, shadedPathPattern);
        }
    }

    public String relocateClass(String clazz) {
        return clazz.replaceFirst(pattern, shadedPattern);
    }

    public String applyToSourceContent(String sourceContent) {
        if (rawString) {
            return sourceContent;
        } else {
            return sourceContent.replaceAll("\\b" + pattern, shadedPattern);
        }
    }
}

File: api/src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/relocation/SimpleRelocator.kt
package com.github.jengelman.gradle.plugins.shadow.relocation

import java.util.regex.Pattern
import org.codehaus.plexus.util.SelectorUtils
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Optional

/**
 * Modified from `org.apache.maven.plugins.shade.relocation.SimpleRelocator.java`
 *
 * @author Jason van Zyl
 * @author Mauro Talevi
 * @author John Engelman
 */
@CacheableRelocator
open class SimpleRelocator @JvmOverloads constructor(
    pattern: String?,
    shadedPattern: String?,
    includes: List<String>?,
    excludes: List<String>?,
    private val _isRawString: Boolean = false,
) : Relocator {
    private val _pattern: String?
    private val _pathPattern: String
    private val _shadedPattern: String?
    private val _shadedPathPattern: String
    private val _includes = mutableSetOf<String>()
    private val _excludes = mutableSetOf<String>()

    init {
        if (_isRawString) {
            _pathPattern = pattern.orEmpty()
            _shadedPathPattern = shadedPattern.orEmpty()
            _pattern = null // not used for raw string relocator
            _shadedPattern = null // not used for raw string relocator
        } else {
            if (pattern == null) {
                _pattern = ""
                _pathPattern = ""
            } else {
                _pattern = pattern.replace('/', '.')
                _pathPattern = pattern.replace('.', '/')
            }
            if (shadedPattern == null) {
                _shadedPattern = "hidden.${_pattern}"
                _shadedPathPattern = "hidden/$_pathPattern"
            } else {
                _shadedPattern = shadedPattern.replace('/', '.')
                _shadedPathPattern = shadedPattern.replace('.', '/')
            }
        }
        _includes += normalizePatterns(includes)
        _excludes += normalizePatterns(excludes)
    }

    @get:Input
    @get:Optional
    open val pattern: String? get() = _pattern

    @get:Input
    open val pathPattern: String get() = _pathPattern

    @get:Input
    @get:Optional
    open val shadedPattern: String? get() = _shadedPattern

    @get:Input
    open val shadedPathPattern: String get() = _shadedPathPattern

    @get:Input
    open val isRawString: Boolean get() = _isRawString

    @get:Input
    open val includes: Set<String> get() = _includes

    @get:Input
    open val excludes: Set<String> get() = _excludes

    open fun include(pattern: String): SimpleRelocator = apply {
        _includes += normalizePatterns(listOf(pattern))
    }

    open fun exclude(pattern: String): SimpleRelocator = apply {
        _excludes += normalizePatterns(listOf(pattern))
    }

    override fun canRelocatePath(path: String): Boolean {
        var adjustPath = path
        if (_isRawString) {
            return Pattern.compile(_pathPattern).matcher(adjustPath).find()
        }

        // If string is too short - no need to perform expensive string operations
        if (adjustPath.length < _pathPattern.length) {
            return false
        }

        if (adjustPath.endsWith(".class")) {
            // Safeguard against strings containing only ".class"
            if (adjustPath.length == 6) {
                return false
            }
            adjustPath = adjustPath.substring(0, adjustPath.length - 6)
        }

        // Allow for annoying option of an extra / on the front of a path. See MSHADE-119 comes from getClass().getResource("/a/b/c.properties").
        val pathStartsWithPattern = if (adjustPath.startsWith("/")) {
            adjustPath.startsWith(_pathPattern, 1)
        } else {
            adjustPath.startsWith(_pathPattern)
        }

        return if (pathStartsWithPattern) isIncluded(adjustPath) && !isExcluded(adjustPath) else false
    }

    override fun canRelocateClass(className: String): Boolean {
        return !_isRawString && !className.contains('/') && canRelocatePath(className.replace('.', '/'))
    }

    override fun relocatePath(context: RelocatePathContext): String {
        val path = context.path
        context.stats.relocate(_pathPattern, _shadedPathPattern)
        return if (_isRawString) {
            path.replace(_pathPattern.toRegex(), _shadedPathPattern)
        } else {
            path.replaceFirst(_pathPattern, _shadedPathPattern)
        }
    }

    override fun relocateClass(context: RelocateClassContext): String {
        val clazz = context.className
        context.stats.relocate(_pathPattern, _shadedPathPattern)
        return clazz.replaceFirst(_pattern.orEmpty(), _shadedPattern.orEmpty())
    }

    override fun applyToSourceContent(sourceContent: String): String {
        return if (_isRawString) {
            sourceContent
        } else {
            sourceContent.replace("\\b$_pattern".toRegex(), _shadedPattern.orEmpty())
        }
    }

    private fun normalizePatterns(patterns: Collection<String>?) = buildSet {
        for (pattern in patterns.orEmpty()) {
            // Regex patterns don't need to be normalized and stay as is
            if (pattern.startsWith(SelectorUtils.REGEX_HANDLER_PREFIX)) {
                add(pattern)
                continue
            }

            val classPattern = pattern.replace('.', '/')
            add(classPattern)

            if (classPattern.endsWith("/*")) {
                val packagePattern = classPattern.substring(0, classPattern.lastIndexOf('/'))
                add(packagePattern)
            }
        }
    }

    private fun isIncluded(path: String): Boolean {
        return _includes.isEmpty() || _includes.any { SelectorUtils.matchPath(it, path, "/", true) }
    }

    private fun isExcluded(path: String): Boolean {
        return _excludes.any { SelectorUtils.matchPath(it, path, "/", true) }
    }
}
